src/luaproc.c

```diff
@@ -424,6 +424,88 @@ static int luaproc_buff_writer( lua_State *L, const void *buff, size_t size,
   return 0;
 }
 
+
+
+static void _copyvalues(lua_State *Lfrom, lua_State *Lto, int type) {
+  const char* str;
+  size_t len;
+  int idx;
+
+  switch(type) {
+  	  case LUA_TSTRING:
+		str = lua_tolstring(Lfrom, -1, &len);
+		lua_pushlstring(Lto, str, len);
+		break;
+  	  case LUA_TNUMBER:
+		copynumber(Lto, Lfrom, -1);
+		break;
+  	  case LUA_TFUNCTION:
+		if(lua_iscfunction(Lfrom, -1)) {
+			lua_pushcfunction(Lto, lua_tocfunction(Lfrom, -1));
+		}
+		break;
+  	  case LUA_TBOOLEAN:
+		lua_pushboolean(Lto, lua_toboolean(Lfrom, -1));
+		break;
+  	  case LUA_TTABLE:
+		lua_newtable(Lto);
+		idx = lua_gettop(Lfrom);
+		lua_pushnil(Lfrom);
+			while(lua_next(Lfrom, idx) != 0) {
+				str = lua_tolstring(Lfrom, -2, &len);
+				if(str == NULL)
+					break;
+				lua_pushlstring(Lto, str, len);
+				_copyvalues(Lfrom, Lto, lua_type(Lfrom, -1));
+				lua_settable(Lto, 5);
+				lua_pop(Lfrom, 1); /* pop value and continue table iteration */
+			}
+		break;
+
+	}
+}
+
+/* copies userdata between lua states, and set their respective metatable (if there's any) */
+static int copyuserdata(lua_State *Lfrom, lua_State *Lto) {
+  void* new_udata;
+  void* ptr_udata;
+  size_t udatasz;
+  size_t len;
+  const char* str;
+  int idx;
+
+  udatasz = lua_rawlen(Lfrom, -1);
+  new_udata = lua_newuserdata(Lto, udatasz);
+  ptr_udata = lua_touserdata(Lfrom, -1);
+  if(ptr_udata == NULL)
+	  return luaL_error(Lfrom, "Userdata it's not complete, see reference manual");
+  memcpy(new_udata, ptr_udata, udatasz);
+  if(lua_getmetatable(Lfrom, -1) == TRUE) {
+	  idx = lua_gettop(Lfrom);
+	  if(lua_getfield(Lfrom, idx, "__name") == LUA_TNIL)
+		  return luaL_error(Lfrom, "Can't get __name from userdata's metatable");
+	  luaL_newmetatable(Lto, lua_tostring(Lfrom, -1));
+	  lua_pop(Lfrom, 1);
+	  lua_pushnil(Lfrom);
+	  while(lua_next(Lfrom, idx) != 0) {
+		  if(lua_type(Lfrom, -2) == LUA_TSTRING) { /* make sure key it's a string */
```

Comment:
"While traversing a table, do not call lua_tolstring directly on a key, unless you know that the key is actually a string. Recall that lua_tolstring may change the value at the given index; this confuses the next call to lua_next. " [1]
[1] http://www.lua.org/manual/5.3/manual.html#lua_next